dbcon/joblist/primitivestep.h

```diff
@@ -1515,21 +1512,31 @@ class TupleBPS : public BatchPrimitive, public TupleDeliveryStep
   boost::shared_ptr<RowGroupDL> deliveryDL;
   uint32_t deliveryIt;
 
-  struct JoinLocalData
+  class JoinLocalData
   {
+  public:    
       JoinLocalData() = delete;
       JoinLocalData(const JoinLocalData&) = delete;
       JoinLocalData(JoinLocalData&&) = delete;
       JoinLocalData& operator=(const JoinLocalData&) = delete;
       JoinLocalData& operator=(JoinLocalData&&) = delete;
       ~JoinLocalData() = default;
 
-      JoinLocalData(rowgroup::RowGroup& primRowGroup, rowgroup::RowGroup& outputRowGroup,
+      JoinLocalData(TupleBPS* pTupleBPS, rowgroup::RowGroup& primRowGroup, rowgroup::RowGroup& outputRowGroup,
                     boost::shared_ptr<funcexp::FuncExpWrapper>& fe2, rowgroup::RowGroup& fe2Output,
                     std::vector<rowgroup::RowGroup>& joinerMatchesRGs, rowgroup::RowGroup& joinFERG,
                     std::vector<boost::shared_ptr<joiner::TupleJoiner>>& tjoiners, uint32_t smallSideCount,
                     bool doJoin);
 
+      friend class TupleBPS;
```

Comment:
If you want to make it proper you need to use getter methods for JoinLocalData attributes are accessed in the code. Friend used here gives me an impression that original idea was partially implemented.

Reply:
Absolutely true. Friend here is to alleviate having to implement a pile of setter/getters.

Reply:
Agree with you on that. The most appropriate way is to add setters/getters but it is up to you whether to stick with this friend approach or to add methods.